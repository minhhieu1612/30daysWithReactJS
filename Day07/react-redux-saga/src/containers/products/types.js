const types = {
  ADD_TO_CART: 'ADD_TO_CART',
  CREATE_PRODUCT: 'CREATE_PRODUCT',
  EDIT_PRODUCT: 'EDIT_PRODUCT',
  DELETE_PRODUCT: 'DELETE_PRODUCT',
  FETCH_REQUEST: 'FETCH_REQUEST',
  FETCH_SUCCESS: 'FETCH_SUCCESS',
  FETCH_FAILURE: 'FETCH_FAILURE',
}

export default types
