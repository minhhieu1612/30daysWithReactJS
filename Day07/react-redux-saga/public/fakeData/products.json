[
  {
    "id": 1,
    "name": "Lillis",
    "price": 14.98,
    "quality": 14
  },
  {
    "id": 2,
    "name": "Linc",
    "price": 9.76,
    "quality": 9
  },
  {
    "id": 3,
    "name": "Goddart",
    "price": 13.99,
    "quality": 22
  },
  {
    "id": 4,
    "name": "Hansiain",
    "price": 3.58,
    "quality": 1
  },
  {
    "id": 5,
    "name": "Keane",
    "price": 7.27,
    "quality": 9
  },
  {
    "id": 6,
    "name": "Kristofor",
    "price": 12.32,
    "quality": 23
  },
  {
    "id": 7,
    "name": "Teddie",
    "price": 12.5,
    "quality": 23
  },
  {
    "id": 8,
    "name": "Hanny",
    "price": 6.49,
    "quality": 3
  },
  {
    "id": 9,
    "name": "Dorris",
    "price": 8.01,
    "quality": 25
  },
  {
    "id": 10,
    "name": "Claudius",
    "price": 7.57,
    "quality": 30
  }
]