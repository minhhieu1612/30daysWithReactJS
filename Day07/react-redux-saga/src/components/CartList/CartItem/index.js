import React, { Component } from 'react'

class CartItem extends Component {
  onDeleteCart = () => {
    let { cart, onDeleteCart } = this.props
    onDeleteCart(cart)
  }

  onEditCartQuality = (increaseQuality) => {
    let { cart, onEditCartQuality } = this.props
    if (increaseQuality === 1 && cart.quality === cart.maxQuality) alert('Cannot get more product from stock')
    else if (increaseQuality === -1 && cart.quality === 1) this.onDeleteCart()
    else onEditCartQuality(cart, increaseQuality)
  }

  render() {
    let { index,
      cart: {
        name,
        price,
        quality
      }
    } = this.props

    return (
      <tr>
        <td style={{ width: '10%' }}>{index + 1}</td>
        <td style={{ width: '40%' }}>{name}</td>
        <td style={{ width: '30%' }}>
          <button
            style={{ marginRight: '10px' }}
            onClick={() => this.onEditCartQuality(-1)}
          > - </button>
          {quality}
          <button
            style={{ marginRight: '10px', marginLeft: '10px' }}
            onClick={() => this.onEditCartQuality(1)}
          > + </button>
        </td>
        <td style={{ width: '10%' }}>${price}</td>
        <td style={{ width: '10%' }}>
          <button
            type="button"
            onClick={this.onDeleteCart}
          >Delete</button>
        </td>
      </tr>
    )
  }
}

export default CartItem
