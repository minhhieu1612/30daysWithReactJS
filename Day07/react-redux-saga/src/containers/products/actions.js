import types from './types'

const createProduct = (product) => ({
  type: types.CREATE_PRODUCT,
  product
})

const editProduct = (product) => ({
  type: types.EDIT_PRODUCT,
  product
})

const deleteProduct = (product) => ({
  type: types.DELETE_PRODUCT,
  product
})

const addToCart = (product) => ({
  type: types.ADD_TO_CART,
  product
})

export {
  createProduct,
  editProduct,
  deleteProduct,
  addToCart,
}