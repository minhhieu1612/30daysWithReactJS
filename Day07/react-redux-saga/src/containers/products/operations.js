import { put, takeLatest, takeEvery, call, delay } from "redux-saga/effects";
import types from "./types";
function* fetchProducts() {
  const products = yield call(
    fetch("http://localhost:3001/fakeData/products.json")
  ).then((response, error) => response);

  yield put({ type: "LOADING" });
  yield delay(500);
  console.log(products);
  // yield put({ type: 'UNLOAD' })
  // yield put({ type: types, products: products })
}

function* watchFetchRequest() {
  // yield put({ type: types.FETCH_REQUEST })
  yield takeLatest(types.FETCH_REQUEST, fetchProducts);
  console.log("has fetch");
}

export default function* rootSaga() {
  yield takeEvery(watchFetchRequest());
}
