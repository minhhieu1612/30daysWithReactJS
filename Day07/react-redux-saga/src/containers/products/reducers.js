import types from './types'
import { findIndex } from '../../helpers'
import products from '../../fakeData/products.json'
import { setProductsStorage, getProductsStorage } from '../../helpers'
const productsStorage = getProductsStorage
const initialState = productsStorage || products

const myReducer = (state = initialState, actions) => {
  switch (actions.type) {
    case types.CREATE_PRODUCT: {
      let product = actions.product
      state.push(product)
      setProductsStorage(state)
    } break
    case types.EDIT_PRODUCT: {
      let product = actions.product
      let index = findIndex(product.id, state)
      if (index > -1) state.splice(index, 1, product)
      setProductsStorage(state)
    } break
    case types.DELETE_PRODUCT: {
      let product = actions.product
      let index = findIndex(product.id, state)
      if (index > -1) state.splice(index, 1)
      setProductsStorage(state)
    } break
    default:
  }

  return [...state]
}

export default myReducer
