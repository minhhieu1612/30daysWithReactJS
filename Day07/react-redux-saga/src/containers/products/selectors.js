/* eslint-disable react/no-typos */
import ListProducts from '../../components/ListProducts'
import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { deleteProduct } from './actions'
import { productEditing } from '../productForm/actions'
import { addToCart } from '../carts/actions'

class ListProductsContainer extends Component {
  render() {
    let { products, onDeleteProduct, onEditingProduct, onAddToCart } = this.props

    return (
      <ListProducts
        products={products}
        onDeleteProduct={onDeleteProduct}
        onEditingProduct={onEditingProduct}
        onAddToCart={onAddToCart}
      />
    )
  }
}

ListProductsContainer.propTypes = {
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      quality: PropTypes.number.isRequired,
    }),
  ).isRequired,
  onDeleteProduct: PropTypes.func.isRequired,
  onEditingProduct: PropTypes.func.isRequired,
  onAddToCart: PropTypes.func.isRequired,
}

const mapStateToProps = state => {
  return {
    products: state.products
  }
}

const mapDispatchToProps = (dispatch, props) => {
  return {
    onDeleteProduct: (product) => dispatch(deleteProduct(product)),
    onEditingProduct: (product) => dispatch(productEditing(product)),
    onAddToCart: (cart) => dispatch(addToCart(cart)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ListProductsContainer)
