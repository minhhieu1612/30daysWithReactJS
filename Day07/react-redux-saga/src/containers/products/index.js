import ListProductsContainer from './selectors'
import operations from './operations'
import reducers from './reducers'

export {
  ListProductsContainer,
  operations,
  reducers
}