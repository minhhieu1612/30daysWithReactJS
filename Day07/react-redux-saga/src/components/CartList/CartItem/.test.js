// import jest from 'jest'
import React from 'react'
import CartItem from '.'
import renderer from 'react-test-renderer'
import products from '../../../fakeData/products.json'

const myMock = jest.fn()

it('testing CartItem component', () => {
  let { id, name, price, quality } = products[0]
  let cart = {
    id, name, price,
    quality: 1,
    maxQuality: quality
  }
  const component = renderer.create(
    <CartItem key={id} index={0} cart={cart} />
  )

  // check on render
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  // manually trigger delete cart
  // tree.props.onDeleteCart
  // const onDeleteCart = myMock.bind(tree.props.onDeleteCart)
  // onDeleteCart()
  // console.log(myMock.mock)
  // tree = component.toJSON()
  // expect(tree).toMatchSnapshot()

  // manually trgigger edit cart quality
  // console.log();
  // tree.children[2].children[2].props.onClick(1)

  // // const onEditCartQuality = myMock.bind(tree.props.onEditCartQuality)
  // // onEditCartQuality(1)
  // tree = component.toJSON()
  // expect(tree).toMatchSnapshot()
});