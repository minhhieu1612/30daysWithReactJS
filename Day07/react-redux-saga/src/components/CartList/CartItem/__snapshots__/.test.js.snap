// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing CartItem component 1`] = `
<tr>
  <td
    style={
      Object {
        "width": "10%",
      }
    }
  >
    1
  </td>
  <td
    style={
      Object {
        "width": "40%",
      }
    }
  >
    Lillis
  </td>
  <td
    style={
      Object {
        "width": "30%",
      }
    }
  >
    <button
      onClick={[Function]}
      style={
        Object {
          "marginRight": "10px",
        }
      }
    >
       - 
    </button>
    1
    <button
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "10px",
          "marginRight": "10px",
        }
      }
    >
       + 
    </button>
  </td>
  <td
    style={
      Object {
        "width": "10%",
      }
    }
  >
    $
    14.98
  </td>
  <td
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Delete
    </button>
  </td>
</tr>
`;
